### PR TITLE
use strict

### PR DESCRIPTION
**Pull request by ocefpaf**
@darribas we can add the extra packages in a another commit. Let's just see if your CIs passes with this change.

### CODE REVIEW

File: gds_py/Dockerfile [hunk 1]
```diff
@@ -8,8 +8,11 @@ ENV LANG=C.UTF-8 LC_ALL=C.UTF-8
 
 #--- Python ---#
 
-RUN conda update --yes conda 
-RUN conda install -c conda-forge -c defaults --quiet --yes \
+RUN conda update conda --yes \
```
Comment (ocefpaf) [inline gds_py/Dockerfile]: The less docker commands you have the less layers and smaller images you'll get. Try to "chain" the commands as much as possible.

File: gds_py/Dockerfile [hunk 2]
```diff
@@ -40,7 +43,12 @@ RUN conda install -c conda-forge -c defaults --quiet --yes \
      'statsmodels' \
      'urbanaccess' \
      'xlrd' \
-     'xlsxwriter'
+     'xlsxwriter' \
+     && conda clean --all --yes --force-pkgs-dirs \
+     && find /opt/conda/ -follow -type f -name '*.a' -delete \
+     && find /opt/conda/ -follow -type f -name '*.pyc' -delete \
+     && find /opt/conda/ -follow -type f -name '*.js.map' -delete \
```
Comment (ocefpaf) [inline gds_py/Dockerfile]: This will ensure a smaller image by removing stuff that you'll probably won't use.

File: gds_py/Dockerfile [hunk 3]
```diff
@@ -40,7 +43,12 @@ RUN conda install -c conda-forge -c defaults --quiet --yes \
...
+     && conda list
 
 # pip commands
 RUN pip install -U --no-deps pysal==2.1.0
```
Comment (ocefpaf) [inline gds_py/Dockerfile]: All but `bambi` are in `conda-forge`, any reason to `pip` install them?

PS: https://github.com/conda-forge/staged-recipes/pull/9053

File: gds_py/Dockerfile [hunk 4]
```diff
@@ -57,7 +65,6 @@ RUN nbdime extensions --enable --user $NB_USER \
  && jupyter labextension update nbdime-jupyterlab
 
 # Clean up
-RUN conda clean -tipsy
```
Comment (ocefpaf) [inline gds_py/Dockerfile]: Moved to the end of the `conda` command to avoid an extra layer. I would join these into a single layer too.

**Comment by darribas** [comment]
Not sure the CI is setup for `gds_py` (actually I don't know how to test containers but would be cool!). Let me pull it and run it locally and see what happens. Will report back!

File: gds_py/Dockerfile [hunk 5]
```diff
@@ -8,8 +8,11 @@ ENV LANG=C.UTF-8 LC_ALL=C.UTF-8
 
 #--- Python ---#
 
-RUN conda update --yes conda 
-RUN conda install -c conda-forge -c defaults --quiet --yes \
+RUN conda update conda --yes \
+    && conda config --add channels conda-forge \
+    && conda config --set channel_priority strict \
+    && conda install --yes --quiet \
+     'python=3.7'
```
Comment (ocefpaf) [inline gds_py/Dockerfile]: I prefer to specify the python version to ensure I know which one will be installed. I can remove this if you prefer not to specify.

**Comment by ocefpaf** [comment]
> Not sure the CI is setup for `gds_py` (actually I don't know how to test containers but would be cool!). Let me pull it and run it locally and see what happens. Will report back!

People usually just build them using external services. I know that some of them even have CI hooks to show the build on GitHub. You can try https://microbadger.com.

**Comment by darribas** [comment]
About to push a couple of typos to your branch

**Comment by darribas** [comment]
There you go, very minor:

https://github.com/ocefpaf/gds_env/pull/1

 I'm away from building machine so can't check on results but will report as soon as I can.

Microbadger looks excellent! Although I haven't been able to find CI testing. Will continue exploring, I think I should be able to set something up on Travis (I'll add it as an issue so I don't forget).

**Comment by darribas** [comment]
Ok I just pushed a minor change (ac23abb), and it fails even before starting :-)

Here's the output:

```docker
PS C:\Users\danie\code\gds_env\gds_py> docker build -t gds_py3 .
Sending build context to Docker daemon  14.85kB
Step 1/14 : FROM jupyter/minimal-notebook:307ad2bb5fce
 ---> ee29ce06a169
Step 2/14 : MAINTAINER Dani Arribas-Bel <D.Arribas-Bel@liverpool.ac.uk>
 ---> Using cache
 ---> b16b6330321f
Step 3/14 : ENV LANG=C.UTF-8 LC_ALL=C.UTF-8
 ---> Using cache
 ---> 70d8043e5cb1
Step 4/14 : RUN conda update conda --yes     && conda config --add channels conda-forge     && conda config --set channel_priority strict     && conda install --yes --quiet      'python=3.7'      'black'      'bokeh'      'contextily==1.0rc2'      'cython'      'dask'      'dask-ml'      'datashader'      'feather-format'      'geopandas'      'hdbscan'      'ipyleaflet'      'ipyparallel'      'ipywidgets'      'mkl-service'      'mplleaflet'      'networkx'      'osmnx'      'palettable'      'pillow'      'pyarrow'      'pymc3'      'pysal'      'qgrid'      'rasterio'      'rasterstats'      'scikit-image'      'scikit-learn'      'seaborn'      'statsmodels'      'xlrd'      'xlsxwriter'      && conda clean --all --yes --force-pkgs-dirs      && find /opt/conda/ -follow -type f -name '*.a' -delete      && find /opt/conda/ -follow -type f -name '*.pyc' -delete      && find /opt/conda/ -follow -type f -name '*.js.map' -delete      && conda list
 ---> Running in 5b8342f218cc
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /opt/conda

  added / updated specs:
    - conda


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    conda-package-handling-1.4.1|           py37_0         262 KB  conda-forge
    jsonschema-3.0.2           |           py37_0          87 KB  conda-forge
    libxml2-2.9.9              |       h8cf2a2b_3         1.3 MB  conda-forge
    mako-1.1.0                 |             py_0          57 KB  conda-forge
    pyrsistent-0.15.4          |   py37h516909a_0          89 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         1.8 MB

The following packages will be REMOVED:

  libiconv-1.15-h516909a_1005

The following packages will be UPDATED:

  conda-package-han~                          1.3.11-py37_0 --> 1.4.1-py37_0
  jsonschema                                   3.0.1-py37_0 --> 3.0.2-py37_0
  libxml2                                  2.9.9-hee79883_2 --> 2.9.9-h8cf2a2b_3
  mako                                          1.0.10-py_0 --> 1.1.0-py_0
  pyrsistent                          0.15.3-py37h516909a_0 --> 0.15.4-py37h516909a_0



Downloading and Extracting Packages
pyrsistent-0.15.4    | 89 KB     | ########## | 100%
libxml2-2.9.9        | 1.3 MB    | ########## | 100%
jsonschema-3.0.2     | 87 KB     | ########## | 100%
conda-package-handli | 262 KB    | ########## | 100%
mako-1.1.0           | 57 KB     | ########## | 100%
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/conda/exceptions.py", line 1062, in __call__
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/conda/cli/main.py", line 84, in _main
    exit_code = do_call(args, p)
  File "/opt/conda/lib/python3.7/site-packages/conda/cli/conda_argparse.py", line 80, in do_call
    module = import_module(relative_mod, __name__.rsplit('.', 1)[0])
  File "/opt/conda/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/opt/conda/lib/python3.7/site-packages/conda/cli/main_install.py", line 8, in <module>
    from .install import install
  File "/opt/conda/lib/python3.7/site-packages/conda/cli/install.py", line 19, in <module>
    from ..core.index import calculate_channel_urls, get_index
  File "/opt/conda/lib/python3.7/site-packages/conda/core/index.py", line 9, in <module>
    from .package_cache_data import PackageCacheData
  File "/opt/conda/lib/python3.7/site-packages/conda/core/package_cache_data.py", line 15, in <module>
    from conda_package_handling.api import InvalidArchiveError
  File "/opt/conda/lib/python3.7/site-packages/conda_package_handling/api.py", line 12, in <module>
    from .tarball import CondaTarBZ2 as _CondaTarBZ2, libarchive_enabled
  File "/opt/conda/lib/python3.7/site-packages/conda_package_handling/tarball.py", line 11, in <module>
    import libarchive
  File "/opt/conda/lib/python3.7/site-packages/libarchive/__init__.py", line 1, in <module>
    from .entry import ArchiveEntry
  File "/opt/conda/lib/python3.7/site-packages/libarchive/entry.py", line 6, in <module>
    from . import ffi
  File "/opt/conda/lib/python3.7/site-packages/libarchive/ffi.py", line 27, in <module>
    libarchive = ctypes.cdll.LoadLibrary(libarchive_path)
  File "/opt/conda/lib/python3.7/ctypes/__init__.py", line 434, in LoadLibrary
    return self._dlltype(name)
  File "/opt/conda/lib/python3.7/ctypes/__init__.py", line 356, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libiconv.so.2: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/bin/conda", line 13, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.7/site-packages/conda/cli/main.py", line 150, in main
    return conda_exception_handler(_main, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/conda/exceptions.py", line 1354, in conda_exception_handler
    return_value = exception_handler(func, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/conda/exceptions.py", line 1065, in __call__
    return self.handle_exception(exc_val, exc_tb)
  File "/opt/conda/lib/python3.7/site-packages/conda/exceptions.py", line 1109, in handle_exception
    return self.handle_unexpected_exception(exc_val, exc_tb)
  File "/opt/conda/lib/python3.7/site-packages/conda/exceptions.py", line 1120, in handle_unexpected_exception
    self.print_unexpected_error_report(error_report)
  File "/opt/conda/lib/python3.7/site-packages/conda/exceptions.py", line 1190, in print_unexpected_error_report
    from .cli.main_info import get_env_vars_str, get_main_info_str
  File "/opt/conda/lib/python3.7/site-packages/conda/cli/main_info.py", line 19, in <module>
    from ..core.index import _supplement_index_with_system
  File "/opt/conda/lib/python3.7/site-packages/conda/core/index.py", line 9, in <module>
    from .package_cache_data import PackageCacheData
  File "/opt/conda/lib/python3.7/site-packages/conda/core/package_cache_data.py", line 15, in <module>
    from conda_package_handling.api import InvalidArchiveError
  File "/opt/conda/lib/python3.7/site-packages/conda_package_handling/api.py", line 12, in <module>
    from .tarball import CondaTarBZ2 as _CondaTarBZ2, libarchive_enabled
  File "/opt/conda/lib/python3.7/site-packages/conda_package_handling/tarball.py", line 11, in <module>
    import libarchive
  File "/opt/conda/lib/python3.7/site-packages/libarchive/__init__.py", line 1, in <module>
    from .entry import ArchiveEntry
  File "/opt/conda/lib/python3.7/site-packages/libarchive/entry.py", line 6, in <module>
    from . import ffi
  File "/opt/conda/lib/python3.7/site-packages/libarchive/ffi.py", line 27, in <module>
    libarchive = ctypes.cdll.LoadLibrary(libarchive_path)
  File "/opt/conda/lib/python3.7/ctypes/__init__.py", line 434, in LoadLibrary
    return self._dlltype(name)
  File "/opt/conda/lib/python3.7/ctypes/__init__.py", line 356, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libiconv.so.2: cannot open shared object file: No such file or directory
The command '/bin/sh -c conda update conda --yes     && conda config --add channels conda-forge     && conda config --set channel_priority strict     && conda install --yes --quiet      'python=3.7'      'black'      'bokeh'      'contextily==1.0rc2'      'cython'      'dask'      'dask-ml'      'datashader'      'feather-format'      'geopandas'      'hdbscan'      'ipyleaflet'      'ipyparallel'      'ipywidgets'      'mkl-service'      'mplleaflet'      'networkx'      'osmnx'      'palettable'      'pillow'      'pyarrow'      'pymc3'      'pysal'      'qgrid'      'rasterio'      'rasterstats'      'scikit-image'      'scikit-learn'      'seaborn'      'statsmodels'      'xlrd'      'xlsxwriter'      && conda clean --all --yes --force-pkgs-dirs      && find /opt/conda/ -follow -type f -name '*.a' -delete      && find /opt/conda/ -follow -type f -name '*.pyc' -delete      && find /opt/conda/ -follow -type f -name '*.js.map' -delete      && conda list' returned a non-zero code: 1
PS C:\Users\danie\code\gds_env\gds_py>
```

**Comment by ocefpaf** [comment]
> Ok I just pushed a minor change (ac23abb), and it fails even before starting :-)

This is b/c of the base image and its frozen dependencies. I recommend starting from a lighter image. Unless you really need the `jupyter/minimal-notebook` one.

**Comment by darribas** [comment]
What's more minimal than that? All the Latex dependencies are useful for me. Any way to "unfreeze" those dependencies?

**Comment by darribas** [comment]
dfc3ad6 doesn't seem to change it, but I think it's because it operates at the `conda install` rather than the `conda update`.

**Comment by ocefpaf** [comment]
> [dfc3ad6](https://github.com/darribas/gds_env/commit/dfc3ad62811a8da2d1f0b3f8d407efa5cfa4dd82) doesn't seem to change it, but I think it's because it operates at the `conda install` rather than the `conda update`.

Nope. It does not. I'm trying to see what we can do here.

> What's more minimal than that? All the Latex dependencies are useful for me. Any way to "unfreeze" those dependencies?

I use the miniconda one. Also playing with it locally... I'll check if we can add the latex stuff in there too.

**Comment by ocefpaf** [comment]
Ignore what I said above. The culprit is this PR: https://github.com/conda-forge/libxml2-feedstock/pull/48

I removed the packages and I'll re-add them once we are ready to use a `libxml2` without `libiconv`.

**Comment by darribas** [comment]
Wahooo!!! This seems to build correctly! Merging right away!!! Will test more tomorrow but this looks like a great start for `GDS-3.0`. THANK you SO very much @ocefpaf!!

**Comment by ocefpaf** [comment]
> Wahooo!!! This seems to build correctly! Merging right away!!! Will test more tomorrow but this looks like a great start for `GDS-3.0`. THANK you SO very much @ocefpaf!!

No problem. These images are usually a great way for us to debug problems with the packages and fix them.

**Pull request merged**